src/backtrace.c

```diff
@@ -0,0 +1,682 @@
+ * so that it's always possible to infer the function a certain address belongs to.
+ * 
+ * Given that not all addresses are saved, it is important to provide accurate
+ * source code references for stack frames that are interrupted by interrupts or 
+ * exceptions; in those cases, the symbolization will simply return the function name
+ * the addresses belongs to, without any source code reference.
+ * 
+ * To see more details on how the symbol table is structured in the ROM, see
+ * #symtable_header_t and the source code of the n64sym tool.
+ * 
+ */
+#include <stdint.h>
+#include <stdalign.h>
+#include <stdlib.h>
+#include <string.h>
+#include "backtrace.h"
+#include "backtrace_internal.h"
+#include "debug.h"
+#include "n64sys.h"
+#include "dma.h"
+#include "utils.h"
+#include "exception.h"
+#include "interrupt.h"
+#include "rompak_internal.h"
+
+/** @brief Enable to debug why a backtrace is wrong */
+#define BACKTRACE_DEBUG 0
+
+/** @brief Function alignment enfored by the compiler (-falign-functions). 
+ * 
+ * @note This must be kept in sync with n64.mk.
+ */
+#define FUNCTION_ALIGNMENT      32
+
+#define MAX_FILE_LEN 120        ///< Maximum length of a file name in a backtrace entry
+#define MAX_FUNC_LEN 120        ///< Maximum length of a function name in a backtrace entry
+
+/** 
+ * @brief Symbol table file header
+ * 
+ * The SYMT file is made of three main tables:
+ * 
+ * * Address table: this is a sequence of 32-bit integers, each representing an address in the ROM.
+ *   The table is sorted in ascending order to allow for binary search. Moreover, the lowest 2 bits
+ *   of each address can store additional information: If bit 0 is set to 1, the address is the start
+ *   of a function. If bit 1 is set to 1, the address is an inline duplicate. In fact, there might be
+ *   multiple symbols at the same address for inlined functions, so we need one entry in this table
+ *   for each entry; all of them will have the same address, and all but the last one will have bit
+ *   1 set to 1.
+ * * Symbol table: this is a sequence of symbol table entries, each representing a symbol. The size
+ *   of this table (in number of entries) is exactly the same as the address table. In fact, each
+ *   address of the address table can be thought of as an external member of this structure; it's
+ *   split externally to allow for efficiency reasons. Each entry stores the function name, 
+ *   the source file name and line number, and the binary offset of the symbol within the containing
+ *   function.
+ * * String table: This tables can be thought as a large buffer holding all the strings needed by all
+ *   symbol entries (function names and file names). Each symbol entry stores a string as an index
+ *   within the symbol table and a length. This allows to reuse the same string (or prefix thereof)
+ *   multiple times. Notice that strings are not null terminated in the string table.
```

Comment:
Do we also optimize for substrings?

Reply:
The file format would allow that, true. 

The limit is on n64sym, because it becomes quadratic to search for a matching substring in an existing table. So in n64ysm we create a hash map with all prefixes (not all possible substrings), and then look that up. 

I think we could improve it adding to the hashmap some special substrings that we feel are more important than others. Assuming a C symbol like `symbol_add`, we could add to the hash map the word `add` (so any word separated by `_`, or even camel case) as it's more likely for a symbol `add` to exist compared to a symbol called `bol_a` (a random substring).

Anyway, I measured that, even using a full substring search, the gain in size is really minimal, as it rarely happens, so I didn't spend too much time on it.

Reply:
Yeah, I don't expect much difference. Just was curious b/c gcc seem to do it for sreing constants. In that case they all anchor from the trailing `\0` so probably it is a much smaller search space